vm/vm.go

```diff
@@ -477,133 +504,154 @@ func parse(
 
 	ctx := &core.Context{
-		GenesisID:        cfg.GenesisID,
-		Registry:         reg,
-		Loader:           loader,
-		PrincipalAccount: account,
-		LayerID:          lid,
-	}
-
-	if account.TemplateAddress != nil {
-		ctx.PrincipalHandler = reg.Get(*account.TemplateAddress)
+		GenesisID:          cfg.GenesisID,
+		Registry:           reg,
+		Loader:             loader,
+		PrincipalAddress:   principal,
+		PrincipalNextNonce: principalAccount.NextNonce,
+		LayerID:            lid,
+		Logger:             logger,
+	}
+
+	// There are three cases to consider:
+	// 1. Principal account does not exist at all (and has no balance). In this case, we can fail
+	// the tx immediately.
+	// 2. Principal account exists, and is spawned. In this case, we use the principal account
+	// template.
+	// 3. Principal account exists as a stub with a nonzero balance, but has not been spawned. In
+	// this case, we assume the tx is a self-spawn for the principal, and check that the calculated
+	// principal matches.
+
+	// NOTE: Athena currently does not allow a tx with principal A to directly call a method on
+	// template B where A != B. That will be handled by "proxied calls", where the target template
+	// is passed not explicitly as part of the tx, but implicitly in the args. This simplifies the
+	// logic here considerably.
+
+	ctx.Header.MaxGas = core.ATHENA_GAS_SPEND + core.ATHENA_GAS_VERIFY
+	if principalAccount.Address == (types.Address{}) {
+		// case 1: principal account does not exist at all
+		return nil, nil, fmt.Errorf("%w: principal account %s does not exist", core.ErrMalformed, principal)
+	} else if principalAccount.TemplateAddress != nil {
+		// case 2: principal account exists and is spawned
+		// attempt to load its template handler
+		// Note: the Wallet template is currently the only supported template, so we could skip this
+		// step and hardcode it here. But this is written in a more future-proof fashion, since we
+		// intend to add support for multiple templates soon.
+		ctx.PrincipalHandler = reg.Get(*principalAccount.TemplateAddress)
 		if ctx.PrincipalHandler == nil {
-			return nil, nil, nil, fmt.Errorf("%w: unknown template %s", core.ErrMalformed, *account.TemplateAddress)
+			return nil, nil, fmt.Errorf("%w: unknown template %s", core.ErrMalformed, principalAccount.TemplateAddress)
 		}
-		ctx.PrincipalTemplate, err = ctx.PrincipalHandler.Load(account.State)
-		if err != nil {
-			return nil, nil, nil, err
+		ctx.Header.TemplateAddress = *principalAccount.TemplateAddress
+	} else {
+		// case 3: principal account exists but is not spawned
+		// go ahead and assume it's a self-spawn for a Wallet template
+		ctx.PrincipalHandler = reg.Get(wallet.TemplateAddress)
+		if ctx.PrincipalHandler == nil {
+			return nil, nil, fmt.Errorf("%w: wallet template missing", core.ErrInternal)
 		}
+		ctx.Header.TemplateAddress = wallet.TemplateAddress
+		ctx.SpawnTx = true
+		ctx.Header.MaxGas = core.ATHENA_GAS_SPAWN + core.ATHENA_GAS_VERIFY
 	}
```

Comment:
How will it support spawning using other templates? Shouldn't the spawn TX also carry the template address?

Reply:
yes, once we have multiple templates, spawn tx should of course include the template address of the template it’s spawning. this is how it works in spacemesh today. for simplicity, for now, I removed the template entirely from tx since there is only one template for now.